모집글, 스크랩 리스트 조회 기능 추가, 수정

yunhozz opened the pull request:
- post 엔티티가 수정됨에 따른 조회에 필요한 관련 dto 추가, 쿼리 수정
- post 상태 조회와 관련된 exception 추가
- 테스트 수정

yunhozz committed:
Remove: 파일 삭제

yunhozz committed:
Fix: post 에 관한 예외 추가, PostRepository DI 추가

- PostAlreadyDeletedException 추가
- postId 로 post 에 대한 조회 추가

yunhozz committed:
Fix: 스크랩한 post 리스트 조회 기능 수정

yunhozz committed:
Test: 엔티티 수정, 조회 수정에 의한 테스트 수정

jjoylee commented on src/main/java/com/justcodeit/moyeo/study/application/scrap/ScrapService.java:
```diff
@@ -15,17 +20,30 @@
+    Post post = postRepository.findById(postId)
+            .orElseThrow(PostCannotFoundException::new);
+
+    if (post.getPostStatus() == PostStatus.DELETE) {
```
Post안에 isDeleted() 메서드로 분리하는 것도 좋을 것 같습니다.

yunhozz replied:
넵 확인했습니다!

yunhozz committed:
Fix: 기술 개수 3개로 제한되게 수정

yunhozz committed:
Refactor: 포스트 삭제 확인 메소드 리팩토링

yunhozz committed:
Fix: static 삭제

yunhozz commented on src/main/java/com/justcodeit/moyeo/study/persistence/repository/scrap/ScrapCustomRepositoryImpl.java:
```diff
@@ -1,35 +1,76 @@
+            .from(postSkill)
+            .join(postSkill.post, post)
+            .join(postSkill.skill, skill)
+            .where(
+                    post.id.in(postIds),
+                    skill.orderNum.in(
+                            JPAExpressions.select(skill.orderNum)
+                                    .from(skill)
+                                    .orderBy(skill.orderNum.asc().nullsLast())
+                                    .limit(3)
```
where 절에서 서브쿼리를 사용할시 limit 구문이 제대로 동작하지 않는 버그가 있다고 합니다. 좀 더 고민해볼게요

yunhozz replied:
limit 을 지우고 대신에 for문을 돌려서 조회된 postSkill이 3개 초과라면 subList 하는 방식으로 수정하였습니다. 하지만, 스킬의 우선도를 null 로 설정할 시, 조회가 제대로 되지 않는 오류가 발생하고 임의의 숫자(ex.999)로 설정해야 잘 됩니다.

mycode01 commented on src/main/java/com/justcodeit/moyeo/study/application/exception/ErrorCode.java:
```diff
@@ -16,9 +16,8 @@ public enum ErrorCode {
 
   NOT_FOUND_USER(404, "U001", "No such User"),
   NOT_FOUND_POST(404, "P001", "No such Post"),
+  POST_ALREADY_DELETED(404, "P002", "Post Already Deleted"),
```
delete는 확인이 필요한 오퍼레이션이 아닙니다.
멱등성을 지키기 위해서라도 없어져야 할 응답같이 느껴지네요.

mycode01 replied:
설명이 좀 더 필요할거같아서 답니다.
사용자는 delete를 할때 해당 항목이 더 이상 필요없기 때문에 delete를 시킬겁니다.
확인이 필요한 오퍼레이션이 아니라는 이야기는 사용자 입장에서, 디비에서 해당 항목이 이미 삭제가 돼었는데 '이미 삭제된 항목입니다' 라는 에러를 보기 위해서 기다리지는 않을거라는 이야깁니다.

yunhozz replied:
저는 이 에러코드를 설정한 이유가 포스트를 삭제했을 때를 의미하는 것이 아니라, 처음에 모집글 리스트를 로드하고 사용자가 스크랩을 하려했을 때, 그 사이 모집글이 삭제되었을 때를 생각해서 만들게 되었습니다.

mycode01 commented on src/main/java/com/justcodeit/moyeo/study/application/scrap/ScrapService.java:
```diff
@@ -15,17 +20,26 @@
+    Post post = postRepository.findById(postId)
+            .orElseThrow(PostCannotFoundException::new);
+
+    isPostDeleted(post);
```
delete 상태인 모집글은 scrap의 대상이 되지 않는다 인가요?

yunhozz replied:
위에와 비슷한 답변이 될 것 같습니다. 아니면 모집글의 삭제 유무와 상관없이 스크랩을 일단 한다음, 나중에 자신이 스크랩한 리스트를 불러왔을 때 삭제된 모집글은 제외하는 식으로 방향을 잡는 것이 더 나을까요..??

mycode01 replied:
아뇨 방향성은 윤호님께서 생각하신게 맞는거같습니다 그냥 모르는 내용이라 물어본거에요
다만 위의 내용과 함께 
삭제상태인 모집글을 불러왔다고 해서 그걸 '이미 삭제된 모집글' 이라고 보여준다 vs '404를 보여준다'의 차이는 좀 있을거같네요.

yunhozz replied:
전자가 더 나을 것 같네요. 204(no content) 에러코드가 적절할까요??

mycode01 commented on src/main/java/com/justcodeit/moyeo/study/application/scrap/ScrapService.java:
```diff
@@ -15,17 +20,26 @@
   @Transactional
   public void deleteScrap(Long scrapId) {
     Scrap scrap = findScrap(scrapId);
+    Post post = postRepository.findById(scrap.getPostId())
```
여기부터 논리에러가 좀 있는거같습니다.
'기존 등록된 scrapId 에게서 post를 찾아와 해당 post가 delete 상태라면 scrap 삭제를 거부한다' 로 보이는데 맞나요?
그리고 전에도 로직이 빠진거같다고 코멘트를 달았지만 
'scrap 삭제 대상이 요청을 건내준 user의 것인지 확인'하는 부분이 빠진거같습니다.

yunhozz replied:
이건 제 논리가 틀린 것 같습니다.
애초에 모집글이 삭제된 스크랩건을 조회하지 않는다면 따로 판별이 필요하지 않을 것이고, 반대로 모집글의 삭제 유무에 상관 없이 스크랩 리스트에 포함이 되어 있다고 해도 이러한 상황과 스크랩 삭제와는 관계가 없을 것 같습니다.
영일님 말씀대로 유저 본인의 스크랩이 맞는지? 를 판별하는 것이 더 나을 것 같습니다. 하지만, 애초에 자신이 스크랩한 리스트를 조회할 때 jwt 토큰으로 판별해서 유저 본인의 고유한 리스트를 받아올 텐데, 그럼 유저의 것인지 따로 판별할 이유가 있을까요..?

mycode01 replied:
애초에 모집글이 삭제된 스크랩건을 조회하지 않는다면 따로 판별이 필요하지 않을 것이고 > 유저는 본인이 스크랩한 모집글이 '삭제상태'가 되었다고해서 스크랩 리스트에서 빠져있다면 당황하지 않을까요?

애초에 자신이 스크랩한 리스트를 조회할 때 jwt 토큰으로 판별해서 유저 본인의 고유한 리스트를 받아올 텐데, 그럼 유저의 것인지 따로 판별할 이유가 있을까요 > scrapId 가 유추 가능할정도로 연속적이기 때문에 악의적 의도를 가진 사용자가 있다면 충분히 접근 가능한 시나리오라고 생각합니다. 그냥 모든 api에 대한 파라메터는 기본적으로 validate가 필요하다고 생각하시면 됩니다. 
이 경우에는 id 를 유추가 힘들게 생성하던지, 아니면 자신의 스크랩인지 확인하는 validation 절차를 만들던지의 방법이 있을거같네요.

yunhozz replied:
아하... 좋은 말씀 감사드립니다. scrap id 값이 랜덤이 아니라서 충분히 악용할 수 있는 시나리오까지 생각을 못했어요..
영일님 말씀대로 자신이 스크랩한 모집글인지 validate 하는 로직이 필요할 것 같습니다. 감사합니다.

mycode01 commented on src/main/java/com/justcodeit/moyeo/study/persistence/repository/scrap/ScrapCustomRepositoryImpl.java:
```diff
@@ -1,35 +1,76 @@
             .join(post).on(scrap.postId.eq(post.id))
-            .where(scrap.userId.eq(userId))
-            .orderBy(scrap.createdDate.desc())
+            .where(scrap.userId.eq(userId), post.id.isNotNull())
```
실제로 돌려본게 아니라 추측해서 이야기 합니다.
post.id is not null 항목을 넣은걸 보면 위의 join이 left join이 아니라서 이렇게 적은거같습니다. 
굳이 필요한 조건은 아닌듯 싶네요.

yunhozz replied:
넵 수정하겠습니다!

mycode01 commented on src/main/java/com/justcodeit/moyeo/study/persistence/repository/scrap/ScrapCustomRepositoryImpl.java:
```diff
@@ -1,35 +1,76 @@
+            .collect(Collectors.toList());
+
+    // 각 포스트마다 스킬 리스트 조회
+    List<PostSkillQueryDto> postSkillDtoList = queryFactory
```
리뷰가 필요할거같습니다.

mycode01 commented on src/test/java/com/justcodeit/moyeo/study/persistence/repository/scrap/ScrapRepositoryTest.java:
```diff
@@ -28,30 +36,63 @@ class ScrapRepositoryTest {
   }
 
   @Test
+  @Commit
```
이 부분은 필요 없을거같습니다.

yunhozz replied:
H2에 테스트하다가 실수로 넣어버렸어요.. 수정하겠습니다

yunhozz committed:
Fix: 리뷰 반영

yunhozz committed:
Fix: post 당 스킬 수 3개 이하로 조회되도록 로직 추가

yunhozz commented on src/main/java/com/justcodeit/moyeo/study/persistence/repository/scrap/ScrapCustomRepositoryImpl.java:
```diff
@@ -1,35 +1,85 @@
+    Map<Long, List<PostSkillQueryDto>> postSkillDtoListMap = postSkillDtoList.stream()
+            .collect(Collectors.groupingBy(PostSkillQueryDto::getPostId));
+
+    postSkillDtoListMap.forEach((postId, dtos) -> {
+        if (dtos.size() > 3) {
+            List<PostSkillQueryDto> subDtos = new ArrayList<>(dtos.subList(0, 3));
+            postSkillDtoListMap.put(postId, subDtos);
+        }
+    });
```
적용되지 않는 limit 구문을 지우고, 대신에 각 postId 마다 조회된 postSkillQueryDto 의 개수가 3개 초과 시 subList 하는 방식으로 로직을 구성해 보았습니다.
하지만, 스킬 우선도를 null 로 설정했을 때 조회가 잘 안됐으며, 임의의 숫자 (ex. 999) 로 설정했을 때는 정상적으로 동작합니다.

yunhozz committed:
Rename: 조회용 dto 폴더명 변경

- scrap -> inquiry

yunhozz committed:
Feat: 모집글, 스크랩 리스트 조회 api 생성

yunhozz committed:
Feat: 스크랩 생성, 삭제시 각각 validation 추가

- 생성시 post 가 이미 삭제된 상태라면 PostAlreadyDeletedException 호출
- 삭제시 해당 스크랩이 유저의 것이 아니라면 ScrapNotAuthorizedException 호출

yunhozz committed:
Fix: 스크랩 삭제 api 에 user token 파라미터 추가

yunhozz committed:
Rename: 패키지 수정에 따른 import 문 수정

yunhozz committed:
Merge branch 'develop' into feature/221018_scrap_inquiry_modify

beetrbgus commented on src/main/java/com/justcodeit/moyeo/study/application/scrap/ScrapService.java:
```diff
@@ -15,17 +21,29 @@
+            .orElseThrow(PostCannotFoundException::new);
+
+    // TODO : http 상태코드? postStatus abnormal?
+    if (post.getPostStatus() == PostStatus.DELETE) {
```
post.getPostStatus() != PostStatus.Normal 이 적당할 것 같아요 보여지는게 Normal 상태인거니까

yunhozz replied:
넵 수정하겠습니다!

beetrbgus commented on src/main/java/com/justcodeit/moyeo/study/interfaces/resource/InquiryController.java:
```diff
@@ -0,0 +1,32 @@
+
+@RestController
+@RequiredArgsConstructor
+public class InquiryController {
```
이거 역할이 뭔가요??
필요 컨트롤러??

yunhozz replied:
자신이 작성한 모집글 리스트, 북마크한 리스트를 조회하는 용도의 컨트롤러를 따로 만들었습니다.

yunhozz committed:
Fix: 예외 추가, 수정 & 관련 로직 추가

- PostAlreadyScrappedException 추가
- 리뷰 반영

yunhozz committed:
Merge remote-tracking branch 'origin/feature/221018_scrap_inquiry_modify' into feature/221018_scrap_inquiry_modify

yunhozz committed:
Fix: 스크랩 존재 여부 로직 수정

yunhozz committed:
Test: 테스트 수정

yunhozz committed:
Test: 테스트 수정

mycode01 commented on src/main/java/com/justcodeit/moyeo/study/application/scrap/ScrapService.java:
```diff
@@ -15,17 +22,34 @@
+    Post post = postRepository.findById(postId)
+            .orElseThrow(PostCannotFoundException::new);
+
+    if (post.getPostStatus() != PostStatus.NORMAL) {
```
아래 if 와 묶어서 별도의 validate 메소드로 빼는 편이 깔끔하고 변경에도 대처하기 쉬울거같습니다.

yunhozz replied:
넵 수정하겠습니다

mycode01 commented on src/main/java/com/justcodeit/moyeo/study/application/scrap/ScrapService.java:
```diff
@@ -15,17 +22,34 @@
+    Scrap scrap = scrapRepository.findByUserIdAndPostId(userId, post.getId())
+            .orElseThrow(ScrapCannotFoundException::new);
+
+    if (!scrap.getUserId().equals(userId)) {
```
이렇게 사용해도 별 문제는 없지만 
scrap을 단순한 데이터 객체만으로 보지말고 자신의 userId와 비교하는 메소드를 제공하는 편이 좀 더 깔끔할거같습니다.

mycode01 commented on src/main/java/com/justcodeit/moyeo/study/interfaces/resource/InquiryController.java:
```diff
@@ -0,0 +1,32 @@
+  private final PostService postService;
+  private final ScrapService scrapService;
+
+  @GetMapping("/posts")
```
라우팅 주소가 좀 애매할수 있을거같습니다.
지금껏 각 컨트롤러가 user, post 를 담당했는데
전혀 다른곳에서 posts가 나와버리기도 하고, 정작 post 쪽에서 get 을 사용해야할때 좀 난감할수 있을거같아요.

yunhozz replied:
흠.. 그렇군요 기존 방식대로 수정하겠습니다!

mycode01 commented on src/main/java/com/justcodeit/moyeo/study/interfaces/resource/ScrapController.java:
```diff
@@ -1,42 +1,33 @@
   @DeleteMapping
-  public ResponseEntity<Void> deleteScrap(@RequestParam String scrapId) {
-    scrapService.deleteScrap(Long.valueOf(scrapId));
+  public ResponseEntity<Void> deleteScrap(@AuthenticationPrincipal UserToken userToken, @RequestParam String postId) {
```
delete 에는 queryparam 보다 pathvariable 이 더 자연스러워 보이지 않을까 합니다.

yunhozz replied:
깜박하고 수정을 안했습니다 감사합니다

mycode01 commented on src/main/java/com/justcodeit/moyeo/study/persistence/repository/querydsl/PostCustomRepositoryImpl.java:
```diff
@@ -51,6 +63,60 @@ public List<Post> findAll(Pageable pageable) {
 
+    @Override
+    public List<PostQueryDto> findPostListByUserId(String userId) {
+        List<PostQueryDto> postDtoList = jpaQueryFactory
```
추상화 레벨을 지키지 않으면 추상화 개념이 없는 곳에서 1000줄짜리 메소드를 수정하게 됩니다.

yunhozz replied:
넵 일단 주먹구구식으로 길게 코드를 짜긴 했는데 저 역시도 너무 길고 복잡하다고 생각이 들었습니다.
추상화 레벨을 고려해서 리팩토링 할 수 있도록 고민해 보겠습니다

yunhozz committed:
Fix: 리뷰 반영

yunhozz committed:
Merge branch 'develop' into feature/221018_scrap_inquiry_modify

yunhozz committed:
Fix: 리뷰 반영

yunhozz committed:
Merge remote-tracking branch 'origin/feature/221018_scrap_inquiry_modify' into feature/221018_scrap_inquiry_modify
